src/crimson/os/seastore/seastore_types.h

```diff
@@ -1374,6 +1449,9 @@ struct record_t {
   record_size_t size;
   mod_time_point_t commit_time;
   record_commit_type_t commit_type;
+  journal_seq_t wouldbe_journal_tail;
+  journal_seq_t wouldbe_dirty_replay_from;
+  journal_seq_t wouldbe_alloc_replay_from;
```

Comment:
I don't really understand this.  Currently, I think we only write down journal_tail in the segment header, so the I suppose the first change here is to write down journal_tail with every record.  I'm not sure that's actually useful, it does reduce work during replay somewhat, but at the cost of expanding the size of the journal header.  The second change is that we distinguish dirty_replay from alloc_replay.  Why is that necessary?

Reply:
> Currently, I think we only write down journal_tail in the segment header, so the I suppose the first change here is to write down journal_tail with every record. I'm not sure that's actually useful, it does reduce work during replay somewhat, but at the cost of expanding the size of the journal header.

Um, the purpose of writing journal_tail to each record is that, I wanted to avoid the scenarios in which paddr allocs/releases that have already applied to the backref tree are replayed. Currently, deltas that should be prevented from replaying are not replayed by judging whether their corresponding extents are live, which seems not easy in case of backref tree. So I record the journal_tail at each record, so when replaying, no outdated paddr allocs/releases(alloc/release deltas) are replayed. I also recorded them in the segment tail.

Seem that I forgot to traverse the non-full journal segment for the exact journal tail at the moment of segment cleaner mount. Will add it:-)

> The second change is that we distinguish dirty_replay from alloc_replay. Why is that necessary?

The reason of this is that dirty extents and dirty alloc/release infos may not be trimmed at the same pace, so we need a way during replay to judge whether a extent modification delta or alloc/release delta should be replayed. Journal_tail only tells that: "from this point on, records contains either valid extent modification deltas or alloc/release deltas", but doesn't specify which one is valid.

Reply:
> Um, the purpose of writing journal_tail to each record is that, I wanted to avoid the scenarios in which paddr allocs/releases that have already been applied to the backref tree are replayed.

I'm trying to understand if it is really true. For example the following allocation events:
1. ---- deltas 2-7 are applied to disk
2. allocate p1 -> l1
3. allocate p2 -> l2
4. allocate p3 -> l3
5. free p2
6. allocate p2 -> l4
7. free p3
8. ---- on-disk backref tree: p1->l1, p2->l4
9. ---- with delta 10-12 committed but not yet applied to disk
10. allocate p3 -> l5
11. free p2
12. allocate p2 -> l6
13. free p1
14. ---- committed backref tree: p2->l6, p2->l5

Seems to me it is OK to apply the already applied deltas for the back-ref tree to rebuild the final view, either from delta 2~7, as long as the replaying is orderred ... is that true?

Maybe the example is over-simplified, or I'm still missing something.

Reply:
Sorry for the delayed response. 

Actually, I think you are right, and I did thought about this, but like you, I couldn't be sure whether it's OK to allow replaying outdated alloc deltas.

And what's more, I thought, if we disallow this scenario, bugs in the code of the backref tree can be identified more easily, for example, any insertion collision are invalid and must be caused by a bug. So I did this. But I'm totally OK if we eventually decide to allow replaying outdated deltas.

Reply:
I think the best way to go is to tolerate replaying outdated alloc deltas.  At first glance, it seems like replaying any suffix should be safe.  Duplicate insertions would still be a bug outside of replay, and we can always check the backref tree against the lba tree to detect bugs.

Reply:
It might be possible to provide a parameter to tolerate collisions during replay to allow outdated deltas, then do strict checks at runtime.

Reply:
yeah, will do

Reply:
@athanatos @cyx1231st When trying to drop this part of code, I encountered the following issue:

- since we don't do writes before osd boot completes, all replayed alloc deltas are cached in Cache::backref_bufs_to_flush, which means that Cache's backref buffers can contain backrefs that already exist in the backref tree. This leads to a problem, that is we are short of methods to exclude backrefs that exist in both the backref tree and the backref buffers when scanning the backref tree, the only way is to search the backref tree for each backref in the backref buffers, which seems very inefficient and could make the osd boot phase very slow. So, I think maybe we should keep `alloc_replay_from` field to avoid replaying outdated alloc deltas, the other two fields can be dropped easily, what do you think? Thanks

Reply:
I'm wondering if this also works:

IIUC only a specific type of transaction/record can trim the journal and update the journal tails, can we add a special delta in the record body to contain the information about the latest journal tails, instead of putting them in every record header?

We can discard the out-dated backref buffers after knowing about the latest journal tails.

Reply:
I'll try that

Reply:
Seems reasonable.